### PR TITLE
Splits url encoding

### PR DESCRIPTION
**Pull request by luoken**
This PR splits up the encoding process when building the URL. Initially when we were using `URI.encode/1` after we built the URL some chars were not encoded namely the "&" character included in the `id_value` e.g. "Seehund, Puma & Co.". Splitting off the `id_value` and encoding that separately using the `URI.encode_www_form/1` properly encodes the values properly

### CODE REVIEW

**Review by fireproofsocks — CHANGES_REQUESTED** [review]
A red flag here is the lengthy comment on a private function -- that says that things are more complicated than they should be, and the comment would probably be unnecessary if there were an appropriate test in place.

At a minimum, you would need a new test to assert that this behavior is now happening.  That would also help communicate the case that this is solving for -- you don't want to hide that example in the PR conversation, you want it **explicitly listed as a test.**

Equally importantly, the way you've wired this unnecessarily adds to the cyclomatic complexity.  I don't think there's a reason not to just run the necessary encoding in all cases and avoid the extra `case` statement.

Thirdly, this should not rely on a www-form encoding because this isn't a form submission.  Granted, sometimes APIs are built to receive the results of form submissions and sometimes encodings are handled haphazardly, but this implementation feels wrong and confuses the issue.

I would implement the function using `URI.encode_query/2` because you are in fact encoding the QUERY (not a form submission).  That would enable you to refactor the `build_url` function to something like the following:

```
  defp build_url(id_key, id_value, lang)
       when (id_key in @allowed_id_keys and is_binary(lang)) or is_atom(lang) do
    {
      :ok,
      "https://#{lang}.wikipedia.org/w/api.php?" <>
        URI.encode_query([
          {:action, "parse"},
          {id_key, id_value},
          {:format, "json"},
          {:redirects, true},
          {:prop,
           "text|langlinks|categories|links|templates|images|externallinks|sections|revid|displaytitle|iwlinks|properties|parsewarnings|headhtml"}
        ])
    }
  end
```

`URI.encode_query/2` can accept any enumerable, so this might be easier to read as a map -- the only gotcha is that we have a VARIABLE key (the `id_key`) so we have to either explicitly ADD that to a map OR use the verbose long-form of a keyword list as I did above.

**Comment by luoken** [comment]
I did not try the encode_query and only tried the www-form function.. I shouldve tried both!
Looking at the difference between the verbose long form vs the map.. If we have to explicitly add the `id_key` and `id_map` into the map  i think i am in favor of the keyword list since it still looks sane.

**Review by fireproofsocks — CHANGES_REQUESTED** [review]
The function looks much better, but the test appears to be making a real live HTTP request.  see comment.

File: test/ex_wikipedia/page/page_test.exs [hunk 1]
```diff
@@ -101,5 +101,10 @@ defmodule ExWikipedia.Page.PageTest do
     test ":error on unsupported languages" do
       assert {:error, _} = Page.fetch(12_345, language: 123)
     end
+
+    test "encodes urls with ampersand" do
+      assert {:ok, %Page{url: "https://en.wikipedia.org/wiki/Seehund,_Puma_%26_Co."}} =
```
[the remark this answers is not in the document]
Comment (luoken) [reply test/ex_wikipedia/page/page_test.exs]: got it!

**Review by everettvody — CHANGES_REQUESTED** [review]
This is actually worse because the test now doesn't actually test anything useful.  

For example, this test also passes:

```
    @tag contents: "33137006.json"
    test "encodes urls with ampersand", %{contents: contents} do
      http_client =
        HTTPClientMock
        |> expect(:get, fn _, _, _ ->
          {:ok,
           %{
             body: contents,
             request: %{
               url: "BURP"
             },
             status_code: 200
           }}
        end)

      assert {:ok,
              %Page{
                url: "https://en.wikipedia.org/wiki/Seehund,_Puma_%26_Co."
              }} =
               Page.fetch("BARF",
                 by: :page,
                 language: "en",
                 http_client: http_client
               )
    end
 ```
 
 You have to figure out how to test that the URL is encoded the way you expect: you should be able to do that by being clever with your mock(s) -- maybe it needs to return an input or something.  It might require a mock for the HTTP client and for the parsing.

Worst case scenario would be to make the `build_url/3` function public so you can actually test it.

Related: the name of the test here is too imprecise.  A better name might be "properly encodes URLs" or something.

**Comment by luoken** [comment]
For now I will make `build_url/3` a public function until I circle back around for

**Comment by fireproofsocks** [comment]
I think having the public function is probably more useful just in case you want to request the URL at a later point, e.g. using some other client or just for human debugging.  Maybe we could even augment the `ExWikipedia` behaviour so it includes both a `fetch/2` and a `url/2` function?  

I mean... otherwise you'd need to really tweak the mocks to something like this:
```
    test "properly encodes url" do
      http_client =
        HTTPClientMock
        |> expect(:get, fn encoded_url, _, _ ->
          {:ok,
           %{
             body: Jason.encode!(%{url: encoded_url}),
             status_code: 200
           }}
        end)

      parser =
        ParserMock
        |> expect(:parse, fn json, _opts ->
          {:ok, json}
        end)

      assert {:ok,
              %{
                url:
                  "https://en.wikipedia.org/w/api.php?action=parse&page=Seehund%2C+Puma+%26+Co.&format=json&redirects=true&prop=text%7Clanglinks%7Ccategories%7Clinks%7Ctemplates%7Cimages%7Cexternallinks%7Csections%7Crevid%7Cdisplaytitle%7Ciwlinks%7Cproperties%7Cparsewarnings%7Cheadhtml"
              }} =
               Page.fetch("Seehund, Puma & Co.",
                 by: :page,
                 http_client: http_client,
                 parser: parser
               )
    end
 ```
 
 and that's pretty hard to follow (even though it passes).
 
 Actually... let's do that: change the name of `build_url` to `url` and add it to the behaviour and always, whenever you have a public function, include a `@doc` block.
